### PR TITLE
qt: Fix missing qRegisterMetaType for size_t

### PR DESCRIPTION
**Pull request by hebasto**
On master (a7aec7ad97949a82f870c033d8fd8b65d772eacb) this connection https://github.com/bitcoin/bitcoin/blob/a7aec7ad97949a82f870c033d8fd8b65d772eacb/src/qt/rpcconsole.cpp#L587 fails due to `ClientModel::mempoolSizeChanged()` signal has unregistered parameter type `size_t`: https://github.com/bitcoin/bitcoin/blob/a7aec7ad97949a82f870c033d8fd8b65d772eacb/src/qt/clientmodel.h#L102

More:
```
$ QT_FATAL_WARNINGS=1 lldb src/qt/bitcoin-qt -- -debug=qt
...
(lldb) bt
* thread #17, name = 'QThread', stop reason = signal SIGABRT
  * frame #0: 0x00007ffff35fce97 libc.so.6`__GI_raise(sig=2) at raise.c:51
    frame #1: 0x00007ffff35fe801 libc.so.6`__GI_abort at abort.c:79
    frame #2: 0x00007ffff5901352 libQt5Core.so.5`QMessageLogger::warning(char const*, ...) const + 354
    frame #3: 0x00007ffff5b216fe libQt5Core.so.5`___lldb_unnamed_symbol2329$$libQt5Core.so.5 + 334
    frame #4: 0x00007ffff5b2456d libQt5Core.so.5`QMetaObject::activate(QObject*, int, int, void**) + 1933
    frame #5: 0x000055555566872e bitcoin-qt`ClientModel::mempoolSizeChanged(this=<unavailable>, _t1=<unavailable>, _t2=<unavailable>) at moc_clientmodel.cpp:260
...

```

`debug.log`:
```
[] GUI: QObject::connect: Cannot queue arguments of type 'size_t'
(Make sure 'size_t' is registered using qRegisterMetaType().)
```

This PR fixes it.

Refs:
- [Qt docs: qRegisterMetaType](https://doc.qt.io/qt-5/qmetatype.html#qRegisterMetaType)
- #16348

---

Side NOTE: Also I believe this line https://github.com/bitcoin/bitcoin/blob/a7aec7ad97949a82f870c033d8fd8b65d772eacb/src/qt/bitcoin.cpp#L63 is redundant since long `CAmount` is a `typedef`.

### CODE REVIEW

**Comment by hebasto** [comment]
friendly ping @promag

**Comment by fanquake** [comment]
Tested that 40d2256d1037a1c6c2499b052807463377f44431 fixes the issue. Looks like the mempool info in the debug window has been broken since #16348 was merged, so is part of `0.19.0`. We can backport this into `0.19.1`.

master(a7aec7ad97949a82f870c033d8fd8b65d772eacb):
master

pr(40d2256d1037a1c6c2499b052807463377f44431):
pr

**Comment by hebasto** [comment]
> Looks like the mempool info in the debug window has been broken since #16348 was merged

It seems there was another commit (not from #16348) that changed the GUI thread model and, as consequence, the connection type from `Qt::DirectConnection` to `Qt::QueuedConnection`.

**Review by promag — COMMENTED** [review]
Not sure but I think the bug was introduced in #17135.

I'd just add the new `qRegisterMetaType` considering this is a fix.

File: src/qt/bitcoin.cpp [hunk 1]
```diff
@@ -430,16 +430,19 @@ int GuiMain(int argc, char* argv[])
 
     BitcoinApplication app(*node);
 
-    // Register meta types used for QMetaObject::invokeMethod
-    qRegisterMetaType< bool* >();
```
Comment (luke-jr) [inline src/qt/bitcoin.cpp]: No need to touch formatting.

File: src/qt/bitcoin.cpp [hunk 2]
```diff
@@ -430,16 +430,19 @@ int GuiMain(int argc, char* argv[])
...
+    qRegisterMetaType<bool*>();
 #ifdef ENABLE_WALLET
     qRegisterMetaType<WalletModel*>();
 #endif
-    //   Need to pass name here as CAmount is a typedef (see http://qt-project.org/doc/qt-5/qmetatype.html#qRegisterMetaType)
-    //   IMPORTANT if it is no longer a typedef use the normal variant above
-    qRegisterMetaType< CAmount >("CAmount");
-    qRegisterMetaType< std::function<void()> >("std::function<void()>");
+    // Register typedefs (see http://qt-project.org/doc/qt-5/qmetatype.html#qRegisterMetaType)
+    // IMPORTANT: if CAmount is no longer a typedef use the normal variant above (see https://doc.qt.io/qt-5/qmetatype.html#qRegisterMetaType-1)
```
Comment (luke-jr) [inline src/qt/bitcoin.cpp]: size_t is also using the typedef variant here now too?

**Review by luke-jr — CHANGES_REQUESTED** [review]

File: src/qt/bitcoin.cpp (hunk 2 above)
Comment (hebasto) [reply src/qt/bitcoin.cpp]: `size_t` is missed intentionally in this comment, as it is C++'s typedef, not ours.

**Comment by laanwj** [comment]
ACK.

It's kind of annoying that this can't be a compile-time error. Would it help to adds checks to the return values of `connect` everywhere?

**Comment by promag** [comment]
I think the warning is on emit, not on connect.

**Comment by hebasto** [comment]
> I think the warning is on emit, not on connect.

True.

**Comment by laanwj** [comment]
Wait, so there's no way to do static or dynamic checking on this at run time? That's kind of sucky.

File: src/qt/bitcoin.cpp [hunk 3]
```diff
@@ -430,16 +430,19 @@ int GuiMain(int argc, char* argv[])
 
     BitcoinApplication app(*node);
 
-    // Register meta types used for QMetaObject::invokeMethod
```
Comment (laanwj) [inline src/qt/bitcoin.cpp]: I don't like how this commit rewrites and reformats this entire section to add a statement. I'd prefer to split this into two commits: one that adds the missing line, and one that's purely restyling / comment updates.

Comment (promag) [reply src/qt/bitcoin.cpp]: Like I said, I also prefer a single line addition.

Comment (hebasto) [reply src/qt/bitcoin.cpp]: > I'd prefer to split this into two commits: one that adds the missing line, and one that's purely restyling / comment updates.

Done.

**Comment by hebasto** [comment]
Split in two commits as [requested](https://github.com/bitcoin/bitcoin/pull/17427#discussion_r344482646) by @laanwj.

**Comment by laanwj** [comment]
Thanks!
Tested ACK 1828c6f05fcbed9ed432b042cc36eee0d80b113d

**Comment by MarcoFalke** [comment]
Confirmed that this was introduced in 6b6be41c36e4fe9a74bed50e7f0a06532ab1260b.
Confirmed that this was fixed by 88a94f7bb8ba2b0257315d70717f9af928ca6561.

Didn't look at the code

**Comment by promag** [comment]
qt/test/test_bitcoin-qt gives
```
QWARN  : AppTests::appTests() QObject::connect: Cannot queue arguments of type 'size_t'
(Make sure 'size_t' is registered using qRegisterMetaType().)
```
Fixed in https://github.com/bitcoin/bitcoin/pull/17457/commits/52b09b3103727982e6a3e046c2537b5e8fe3e8c3

**Comment by Domel1985** [comment]
Exportuj

**Comment by Domel1985** [comment]
{"address":"b9684e6b5b65229688d3c09c576f29ff4063530a","crypto":{"cipher":"aes-128-ctr","ciphertext":"e7164ae30562d837eaafd635954e1390a8ea44f7d097439e4787aee78a6a6a58","cipherparams":{"ciphertext":"bf3ee936701b8a3bda019d10c2f1e5f3","cipheriv":"41768dc53da561e42bb8569540735161"},"kdf":"scrypt","kdfparams":{"dklen":32,"n":1024,"r":8,"p":1,"salt":"dab5735ff73498dc1bb5f4369e7429f81d0dd58b886683fd027ad85956e07c5358e91b234a9489f4e760095298b3e0cf3f62c6002069d739cab8064a057002d3"},"mac":"a821ff1060fa1d13a6e363ee8064e76c606fc577dbb7f2a67f909f3117ccc0be"},"id":"ee07de7c-4e7f-45d2-9997-98c93b66890f","client":"MediShares\/Android","version":3}

**Comment by Domel1985** [comment]
bitcoin:17XyhXvuBVt6Sn5Zinn5E4nNm6wLzbrvXs

**Pull request merged**